### PR TITLE
⚠️ new(403): add k3s control plane type

### LINKED ISSUE #347: LFX Mentorship Term 2 : Extending KubeFlex with a new type of Control Plane
#  Building a Marketplace for KubeStellar Plugins

## Description

KubeFlex is a flexible and scalable platform for running lightweight Kubernetes control plane APIs to support specific use-cases in cloud and edge computing environments. It supports various kinds of control planes, such as: vcluster, ocm, host, etc. This project aims to extend KubeFlex to support a new type of control plane that provides the full components of a control plane in a typical Kubernetes cluster (e.g., API Server, Scheduler, Controllar Manager, etc.). This new control plane will be based on K3s and it will allow KubeFlex to support new use-case such as multi-tenant scenarios.  

## Expected Outcome
A new type of KubeFlex provided control plane based on k3s

## Recommended Skills

Golang, Kubernetes, K3s

## Mentor(s)

- Paolo Dettori: (@pdettori, dettori@us.ibm.com)
- Braulio Dumba: (@dumb0002, Braulio.Dumba@ibm.com)

### PR DESCRIPTION
**Pull request by rxinui**
## Summary

- fix: separate interface for PostCreateHookReconciler
- new: add apiserver, reconciler and service
- ref: refactoring and improvement
- ref: define ControlPlaneReconciler interface
- fix: kflex create --type=k3s finishes execution (#509)
- fix: kubeconfig for user is updated with k3s values
- new: add k3s token as value within k3s-config secret (#513)
- fix: update status Ready=True when k3s server is ready (#517)
- test: add e2e test for k3s controlplane

## Notes

This PR implements the minimum viable product allowing kubeflex to create k3s controlplane.
Follow-ups PR will be necessary to fully cover the edge cases on:

- The reconciliation loop for k3s 
- E2E tests
- Documentation

Reviews on https://github.com/kubestellar/kubeflex/pull/520 will be subject to new issues, rather than blocking this PR and the development of kubeflex, ensuring materials for (new) contributors.

## Related issue(s)

Fixes #347 

## Related issue(s)

Fixes #

### CODE REVIEW

**Comment by rxinui** [comment]
@pdettori any idea why the external-reconciler fails?

File: internal/controller/controlplane_controller.go [hunk 1]
```diff
@@ -161,7 +162,11 @@ func (r *ControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	case tenancyv1alpha1.ControlPlaneTypeHost:
 		reconciler = host.New(r.Client, r.Scheme, r.Version, r.ClientSet, r.DynamicClient, r.EventRecorder)
 	case tenancyv1alpha1.ControlPlaneTypeExternal:
-		reconciler = external.New(r.Client, r.Scheme, r.Version, r.ClientSet, r.DynamicClient, r.EventRecorder)
+		reconciler := external.New(r.Client, r.Scheme, r.Version, r.ClientSet, r.DynamicClient, r.EventRecorder)
+		return reconciler.Reconcile(ctx, hcp)
```
Comment (pdettori) [inline internal/controller/controlplane_controller.go:166]: There is some change in the way the reconciler for external cluster is now created and invoked - could this be the cause of the failing e2e test for external CP ?

File: internal/controller/controlplane_controller.go [hunk 2]
```diff
@@ -201,17 +206,24 @@ func (r *ControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.Request
...
-		}
-
-		// Refresh hcp object after PCH processing
-		if err := r.Get(ctx, client.ObjectKey{Name: hcp.Name}, hcp); err != nil {
-			log.Error(err, "Failed to refresh ControlPlane after hook processing")
-			return ctrl.Result{}, err
+		switch pchReconciler := reconciler.(type) {
+			case shared.PostCreateHookReconciler:
+			// Reconciler that supports PostCreateHook
+				if err := pchReconciler.ReconcileUpdatePostCreateHook(ctx, hcp); err != nil {
```
Comment (pdettori) [inline internal/controller/controlplane_controller.go:212]: The external cluster type does support post-create hooks. Indeed, the broken test is trying to run a post create hook which does not run because the new code for the external cluster never reaches this point, so I think this is the source of the failing e2e test

Comment (rxinui) [reply internal/controller/controlplane_controller.go:210]: Well spotted, some changes from the rebase has been migrated properly by myself. Thank you @pdettori 👍
